core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -24,13 +24,28 @@ import java.nio.file.Files
 import kafka.common.{KafkaException, TopicAndPartition}
 import kafka.utils.{Logging, nonthreadsafe}
 import org.apache.kafka.common.TopicPartition
-import org.apache.kafka.common.errors.{DuplicateSequenceNumberException, InvalidSequenceNumberException, ProducerFencedException}
+import org.apache.kafka.common.errors.{DuplicateSequenceNumberException, OutOfOrderSequenceException, ProducerFencedException}
 import org.apache.kafka.common.protocol.types._
+import org.apache.kafka.common.record.LogEntry
 import org.apache.kafka.common.utils.Utils
 
 import scala.collection.mutable
 
-private[log] case class PidEntry(seq: Int, epoch: Short, offset: Long)
+case class PidEntry(lastSeq: Int,
+                    epoch: Short,
+                    lastOffset: Long,
+                    numRecords: Int,
+                    timestamp: Long
+                   ) {
+  def firstSeq(): Int = {
+    lastSeq - numRecords + 1
+  }
+
+  def firstOffset(): Long = {
+    lastOffset - numRecords + 1
+  }
+}
+private[log] case class PidEntryTuple(firstEntry: PidEntry, lastEntry: PidEntry)
```

Comment:
Will change to `PidEntryRange`. But I would prefer leaving this as `firstEntry` and `lastEntry`. They read better in actual usage. These names were `first` and `last`, but they made the validation code harder to read, IMO.